File: javascript/bird-watcher/bird-watcher.js
// @ts-check
//
// The line above enables type checking for this file. Various IDEs interpret
// the @ts-check directive. It will give you helpful autocompletion when
// implementing this exercise.

/**
 * Number of days in a week.
 */
const DAYS_PER_WEEK = 7;

/**
 * Calculates the total bird count.
 *
 * @param {number[]} birdsPerDay
 * @returns {number} total bird count
 */
export function totalBirdCount(birdsPerDay) {
  let count = 0;
  
  for (let day = 0; day < birdsPerDay.length; day++)
    count += birdsPerDay[day];

  return count;
}

/**
 * Calculates the total number of birds seen in a specific week.
 *
 * @param {number[]} birdsPerDay
 * @param {number} week
 * @returns {number} birds counted in the given week
 */
export function birdsInWeek(birdsPerDay, week) {
  let count = 0;
  let startDay = (week - 1) * DAYS_PER_WEEK;
  let endDay = week * DAYS_PER_WEEK;
  
  for (let day = startDay; day < endDay; day++)
    count += birdsPerDay[day];

  return count;
}

/**
 * Fixes the counting mistake by increasing the bird count
 * by one for every second day.
 *
 * @param {number[]} birdsPerDay
 * @returns {number[]} corrected bird count data
 */
export function fixBirdCountLog(birdsPerDay) {
  for (let day = 0; day < birdsPerDay.length; day += 2)
    birdsPerDay[day]++;
  
  return birdsPerDay;
}

File: go/bird-watcher/bird_watcher.go
package birdwatcher

// TotalBirdCount return the total bird count by summing
// the individual day's counts.
func TotalBirdCount(birdsPerDay []int) int {
	total := 0

	for _, v := range birdsPerDay {
		total += v
	}

	return total
}

// BirdsInWeek returns the total bird count by summing
// only the items belonging to the given week.
func BirdsInWeek(birdsPerDay []int, week int) int {
	total := 0

	last_day := week * 7

	for day := last_day - 7; day < last_day; day++ {
		total += birdsPerDay[day]
	}

	return total
}

// FixBirdCountLog returns the bird counts after correcting
// the bird counts for alternate days.
func FixBirdCountLog(birdsPerDay []int) []int {
	for day := 0; day < len(birdsPerDay); day += 2 {
		birdsPerDay[day]++
	}

	return birdsPerDay
}
